### PR TITLE
Adjusting pattern-syntax.md with recent features

### PR DESCRIPTION
**Pull request by aryx**
This diff mostly reorganize and fix what was there already.
Next commit will document some new features like /$XXX/.

### CODE REVIEW

File: docs/writing-rules/pattern-syntax.md [hunk 1]
```diff
@@ -473,16 +493,43 @@ Many programmers don't really see the difference between `foo()` and `foo();`. T
...
 ```
 
-```text
-pattern: if $CONDITION:
+A complete expression is needed (like `1 + $X`)
+
+## Partial statements
+
+Partial statements are partially supported. For example,
+you can just match the header of a conditional with `if ($E)`,
```
Comment (IagoAbal) [inline docs/writing-rules/pattern-syntax.md:509]: I am not fully aware about where this is particularly useful. When is this preferred to `if ($E) ...` ?

Comment (IagoAbal) [reply docs/writing-rules/pattern-syntax.md:509]: Is this only because you want to match that fragment of the code and nothing else?

Comment (aryx) [reply docs/writing-rules/pattern-syntax.md:509]: Can be useful when used in a pattern-inside, to restrict the context in which to search code

Comment (aryx) [reply docs/writing-rules/pattern-syntax.md:509]: I can probaly add that as a comment

File: docs/writing-rules/pattern-syntax.md [hunk 2]
```diff
@@ -473,16 +493,43 @@ Many programmers don't really see the difference between `foo()` and `foo();`. T
...
+function `foo`. In the same way, you can just match a class header
+(e.g., with `class $A`).
+
+# Deprecated features
+
+## String matching
+
+!!! warning
+    String matching has been deprecated. You should use 
+	[metavariable regexp](rule-syntax.md#metavariable-regex) instead.
```
Comment (IagoAbal) [inline docs/writing-rules/pattern-syntax.md:525]:
```
- [metavariable regexp](rule-syntax.md#metavariable-regex)
+ [`metavariable-regex`](rule-syntax.md#metavariable-regex)
```

**Review by IagoAbal — APPROVED** [review]
I am missing the `function ...` feature that matches both function definitions (stamtents) and lambda expressions. But maybe that goes into a different PR?

**Review by ievans — APPROVED** [review]
🙏 ty for updating the docs!

**Pull request merged**